### PR TITLE
added concept and pipeline definition docs

### PR DESCRIPTION
**Pull request by bergos**
This PR adds documentation about the basic concept, internals (stream stuff), pipeline definitions (steps, variables) and how JS code can be used in the pipeline.

### CODE REVIEW

**Review by vhf — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,5 +1,213 @@
 # barnard59
 
+`barnard59` runs pipelines which are described in RDF in the [pipeline](https://pipeline.described.at/) and [code](https://code.described.at/) ontology.
+
+## Concept
+
+A barnard59 pipeline consists of one or more steps where each step returns a [Stream](https://nodejs.org/api/stream.html).
+The streams of all steps are combined via [.pipe](https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options).
```
Comment (tpluscode) [inline README.md]:
```suggestion
The streams of all steps are combined via [`.pipe`](https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options).
```

File: README.md [hunk 2]
```diff
@@ -1,5 +1,213 @@
...
+<#pipeline> a p:Pipeline, p:Writable, p:Readable .
+```
+
+**Note:** The type of the streams could be also detected during runtime or based on metadata for the streams, but that would required deeper understanding of the pipeline and steps structures.
+To simplify handling of pipeline definitions without runtime data, the stream types are attached to the pipeline definition and errors are thrown in the runtime if they don't match. 
+
+### Step
+
+A pipeline consists of one or more steps, which are actually factories which return streams.
+Each step is linked with the previous one via [.pipe](https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options).
```
Comment (tpluscode) [inline README.md]:
```suggestion
Each step is linked with the previous one via [`.pipe`](https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options).
```

File: README.md [hunk 3]
```diff
@@ -1,5 +1,213 @@
...
+  code:arguments [
+    p:name "url";
+    p:value "http://example.org/"
+  ], [
+    p:name "method";
+    p:value "POST"
+  ] .
+```
+
+The function in `customStreamCode.js` will be called with the following object:
```
Comment (tpluscode) [inline README.md]: I would make it clear somewhere in this section that the implementation must export function, which takes loaded `code:arguments` and returns a `Stream`.

File: README.md [hunk 4]
```diff
@@ -1,5 +1,213 @@
...
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <node:barnard59-base#map>
+  ] ;
+  code:arguments ("""row => {
+    // replace all ö umlauts with oe
+    return row.split('ö').join('oe')
+  }"""^^code:EcmaScript) .
+```
+
+#### EcmaScript Template Literal
```
Comment (tpluscode) [inline README.md:199]: This section kind of repeats `Using Variables in Template Literals`

**Review by tpluscode — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (bergos) [reply README.md:199]: I know. Made the structure, then filled it. Somehow it should be there as separate topic. Maybe we fill it in the future with more stuff how the template literals can be useful and point from the variable section to this one? Maybe even reduce the stuff in the variable section?

Comment (tpluscode) [reply README.md:199]: m'kay. let's keep it for now.

**Review by tpluscode — APPROVED** [review]

**Pull request merged**